### PR TITLE
USDC: @celo/connect + web3js demo

### PR DESCRIPTION
**Pull request by arthurgousset**
Adds CIP64 transaction demo using: 
- web3js
- @celo/connect (with @celo/wallet-local)

## Related

- https://github.com/celo-org/developer-tooling/issues/128

### CODE REVIEW

**Comment by arthurgousset** [comment]
copy/paste insight from @aaronmgdr 

> ok so here is the issue.
> 
>  in order to sign a cip64 tx a Wallet that is a subclass of WalletBase must be used as that is where the actual code for rlp encode and sign is. in the setup you have its using the wallet on web3. which only knows eip1559
> 
> this line https://github.com/celo-org/developer-tooling/blob/679ef0c607500a69f2d1d684740632202ab79ad2/packages/sdk/connect/src/connection.ts#L72
> 
> if it were not commented out would default the wallet on connection to a LocalWallet.  instead of a ReadOnly Wallet. we should either uncomment that or make it so calling connection.addAccount creates an LocalWallet instance (which would make more sense to me)

**Comment by arthurgousset** [comment]
I'm wondering what purpose the `ReadOnlyWallet` interface serves.

```ts
export interface ReadOnlyWallet {
    getAccounts: () => Address[];
    removeAccount: (address: Address) => void;
    hasAccount: (address?: Address) => boolean;
    signTransaction: (txParams: CeloTx) => Promise<EncodedTransaction>;
    signTypedData: (address: Address, typedData: EIP712TypedData) => Promise<string>;
    signPersonalMessage: (address: Address, data: string) => Promise<string>;
    decrypt: (address: Address, ciphertext: Buffer) => Promise<Buffer>;
    computeSharedSecret: (address: Address, publicKey: string) => Promise<Buffer>;
}
```

Source: connect > [wallet.ts](https://github.com/celo-org/developer-tooling/blob/a884b24a242829a9425580e0b75f40526228838f/packages/sdk/connect/src/wallet.ts#L4)

Why is it called "ReadOnly" if you can sign a transaction with a class that implements it?

**Comment by arthurgousset** [comment]
Wondering if `LocalWallet()` is a web3.js specific concept. 

> Local wallets are an in-memory [wallet](https://docs.web3js.org/api/web3-eth-accounts/class/Wallet/) that can hold multiple accounts. Wallets are a convenient way to sign and send transactions in web3.js

Source: https://docs.web3js.org/guides/wallet/local_wallet/

It's not obvious where web3.js stops and Celo-specific constructs start (e.g. `ReadOnlyWallet`)

**Comment by aaronmgdr** [comment]
> Wondering if `LocalWallet()` is a web3.js specific concept.
> 
> > Local wallets are an in-memory [wallet](https://docs.web3js.org/api/web3-eth-accounts/class/Wallet/) that can hold multiple accounts. Wallets are a convenient way to sign and send transactions in web3.js
> 
> Source: https://docs.web3js.org/guides/wallet/local_wallet/
> 
> It's not obvious where web3.js stops and Celo-specific constructs start (e.g. `ReadOnlyWallet`).

while web3 has wallets. you must use the celo wallets to get the signing happening -- see the wallets directory in the sdks folder of developertooling repo.

**Comment by aaronmgdr** [comment]
> Also curious how _wallet and LocalWallet() are defined.

_wallet comes from the params above

LocalWallet would be imported from @celo/wallet-local

**Comment by aaronmgdr** [comment]
> Why is it called "ReadOnly" if you can sign a transaction with a class that implements it?

yeah i cant answer that one. seems like it should be have a better name

**Comment by arthurgousset** [comment]
> > Also curious how _wallet and LocalWallet() are defined.
> 
> _wallet i think comes from the params above (which is now just wallet)
> 
> LocalWallet would be imported from @celo/wallet-local

Thanks, this helped a lot. I used `LocalWallet` in https://github.com/celo-org/feecurrency/pull/3/commits/c9b9e3f727a3b96fd561e64d3c032d298030ad34. 

At the moment, the script is stuck and doesn't timeout or complete. I'll debug that next

File: web3.ts [hunk 1]
```diff
@@ -32,6 +31,11 @@ async function erc20Transfer() {
...
     );
 
+    // Get the sender's address
+    const sender = celoWallet.getAccounts()[0];
+    // Debugging at the moment
+    console.log(`Sender address: ${sender}`)
+
     const transactionReceipt = await connection
         .sendTransaction({
             from: sender,
```
Comment (aaronmgdr) [inline web3.ts:41]: from will be set by default so not strictly needed here

Comment (arthurgousset) [reply web3.ts:41]: Ah, good to know. Let me try without `from`.

Comment (arthurgousset) [reply web3.ts:41]: Hmm when I remove `from` I get a zero address error it seems

```sh
/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3/node_modules/web3-core/src/web3_request_manager.ts:432
                if (error?.message.includes('revert')) throw new ContractExecutionError(error);
                                               ^
ContractExecutionError: Error happened while trying to execute a function inside a smart contract
    at Function._isReverted (/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3/node_modules/web3-core/src/web3_request_manager.ts:432:48)
    at Web3RequestManager._processJsonRpcResponse (/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3/node_modules/web3-core/src/web3_request_manager.ts:373:23)
    at /Users/arthur/Documents/celo-org/feecurrency/node_modules/web3/node_modules/web3-core/src/web3_request_manager.ts:282:13
    at decoratedCallback (/Users/arthur/Documents/celo-org/feecurrency/node_modules/@celo/connect/lib/utils/rpc-caller.js:111:13)
    at /Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-types/src/web3_base_provider.ts:205:5
    at processTicksAndRejections (node:internal/process/task_queues:95:5) {
  innerError: Eip838ExecutionError: execution reverted: ERC20: transfer from the zero address
      at decoratedCallback (/Users/arthur/Documents/celo-org/feecurrency/node_modules/@celo/connect/lib/utils/rpc-caller.js:109:23)
      at /Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-types/src/web3_base_provider.ts:205:5
      at processTicksAndRejections (node:internal/process/task_queues:95:5) {
    innerError: undefined,
    code: undefined,
    receipt: undefined,
    data: undefined
  },
  code: 310,
  receipt: undefined
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Comment (aaronmgdr) [reply web3.ts:41]: oh wierd. its not as resilient as i thought. it seems config.from is sometimes undefined

**Comment by arthurgousset** [comment]
(for future reference)
Also noting that `type: 'cip64'` in the `Connection.sendTransaction` arguments did not accept a `string`

`Cannot convert string to Uint8Array. only supports 0x-prefixed hex strings and this string was given: cip64`

```diff
const transactionReceipt = await connection
        .sendTransaction({
            from: sender,
            to: CONTRACT_ADDRESS,
            gas: 51925, // TODO: implement gas estimation
            maxPriorityFeePerGas: web3.utils.toWei("10", "gwei"),
            maxFeePerGas: web3.utils.toWei("10", "gwei"),
            feeCurrency: "0x874069Fa1Eb16D44d622F2e0Ca25eeA172369bC1", // cUSD fee currency
            data: transactionObject.encodeABI(),
+            type: 'cip64'
        })
```

Error:

```sh
$ yarn ts-node web3.ts 
yarn run v1.22.19
$ /Users/arthur/Documents/celo-org/feecurrency/node_modules/.bin/ts-node web3.ts
Initiating fee currency transaction...
/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-eth-accounts/src/common/utils.ts:344
                        throw new Error(
         ^
Error: Cannot convert string to Uint8Array. only supports 0x-prefixed hex strings and this string was given: cip64
    at toUint8Array (/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-eth-accounts/src/common/utils.ts:344:10)
    at Function.typeToInt (/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-eth-accounts/src/tx/transactionFactory.ts:43:48)
    at Function.fromTxData (/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-eth-accounts/src/tx/transactionFactory.ts:68:37)
    at /Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-eth/src/utils/prepare_transaction_for_signing.ts:145:28
    at Generator.next (<anonymous>)
    at fulfilled (/Users/arthur/Documents/celo-org/feecurrency/node_modules/web3-eth/lib/commonjs/utils/prepare_transaction_for_signing.js:21:58)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Might be a different reason, but keeping track of this here in case.

**Comment by arthurgousset** [comment]
Hint: Learnt that prepending a Node.js command with `DEBUG=*` executed the Node.js script with verbose output in the console. 

For example: 

```sh
$ DEBUG=* yarn ts-node web3.ts
```

**Comment by arthurgousset** [comment]
With the updated script, `@celo/connect` + `web3` (incl. `@celo/wallet-local`) successfully makes a CIP64 transaction. 
For example: https://alfajores.celoscan.io/tx/0x0cbac87c455a0cf9a1cb6bd340f1332d45eb767a570e901d72b386fb58790f79

**Pull request merged**